ops/model.py

```diff
@@ -3579,6 +3580,16 @@ def secret_info_get(
         id = list(info_dicts)[0]  # Juju returns dict of {secret_id: {info}}
         return SecretInfo.from_dict(id, typing.cast(Dict[str, Any], info_dicts[id]))
 
+    @contextlib.contextmanager
+    def _secrets_in_temp_files(self, content: Dict[str, str]) -> Generator[List[str], None, None]:
+        rv: List[str] = []
```

Comment:
I battled this too, because what it does is it sets things up and "returns arguments where temp file names are interpolated and the temp files contain individual secret data" and found it quite hard to compress into a meaningful function name.

Do you have a better idea? Maybe Ben?

Reply:
Maybe `result`?

However, I think that defining a context manager (which is a generator) with `ExitStack` and `enter_context` is all a bit clever, and kind of took me a while to get my head around. I think it would be simpler and clearer with old-fashioned imperative code. Compare the complexity and verbosity of the two versions:

```python
# CONTEXT MANAGER VERSION:
@contextlib.contextmanager
def _secrets_in_temp_files(self, content: Dict[str, str]) -> Generator[List[str], None, None]:
    rv: List[str] = []
    with contextlib.ExitStack() as temp_files:
        for k, v in content.items():
            temp = temp_files.enter_context(tempfile.NamedTemporaryFile(buffering=0))
            temp.write(v.encode('utf-8'))
            rv.append(f'{k}#file={temp.name}')
        yield rv
with self._secrets_in_temp_files(content or {}) as secret_args:
    self._run_for_secret('secret-set', *args, *secret_args)

# NON-CLEVER VERSION:
files = []
try:
    for k, v in content.items():
        f = tempfile.NamedTemporaryFile(buffering=0)
        files.append(f)
        f.write(v.encode('utf-8'))
        args.append(f'{k}#file={f.name}')
    self._run_for_secret('secret-set', *args)
finally:
    for f in files:
        f.close()
```

What do you think? Same number of lines of code, but the first is about twice as many characters and uses several advanced concepts (unnecessarily I think).

Could factor it into a function if need be:

```python
def _run_with_secrets(command, args, content): ...

_run_with_secrets('secret-set', args, content)
```

Reply:
Separately, I wonder if `secret_add` should use `_run_for_secret` rather than bare `_run` (to handle the `SecretNotFoundError`)?

Reply:
> Separately, I wonder if `secret_add` should use `_run_for_secret` rather than bare `_run` (to handle the `SecretNotFoundError`)?

How would you get a "not found" error when *adding* a secret?

Reply:
Actually, I just read your PR description and realized it's almost certainly better to close each file after writing to it, and then delete them at the end. So maybe something more like this (untested):

```python
paths = []
try:
    for k, v in content.items():
        with tempfile.NamedTemporaryFile(encoding='utf-8', delete=False) as f:
            f.write(v)
            paths.append(f.name)
            args.append(f'{k}#file={f.name}')
    self._run_for_secret('secret-set', *args)
finally:
    for p in paths:
        os.remove(p)
```

Reply:
> How would you get a "not found" error when adding a secret?

Oops, good point, no wonder it's only `_run`. :-)